readthedocs/proxito/views/serve.py

```diff
@@ -274,6 +286,172 @@ def _get_canonical_redirect_type(self, request):
+                        filename=path,
+                        redirect_type=redirect_type,
+                        is_external_version=unresolved_domain.is_from_external_domain,
+                    )
+                except InfiniteRedirectException:
+                    # Don't redirect in this case, since it would break things.
+                    # Disable caching, since we failed to redirect.
+                    self.cache_request = False
+
+            # If the path is not empty, the path doesn't resolve to a proper file.
+            if exc.path != "/":
+                raise Http404
+
+            # When the path is empty, the project didn't have an explicit version,
+            # so we need to redirect to the default version.
+            # This is `/ -> /en/latest/` or
+            # `/projects/subproject/ -> /projects/subproject/en/latest/`.
+
+            # A system redirect can be cached if we don't have information
+            # about the version, since the final URL will check for authz.
+            if self._is_cache_enabled(project):
+                self.cache_request = True
+
+            return self.system_redirect(
+                request=request,
+                final_project=project,
+                version_slug=version_slug,
+                filename=exc.path,
+                is_external_version=unresolved_domain.is_from_external_domain,
+            )
+
+        project = unresolved.project
+        version = unresolved.version
+        filename = unresolved.filename
+
+        # TODO: find a better way to pass this to the middleware.
+        request.path_project_slug = project.slug
+        request.path_version_slug = version.slug
+
+        if not version.active:
+            log.warning("Version is not active.")
+            raise Http404("Version is not active.")
+
+        if self._is_cache_enabled(project) and not version.is_private:
+            # All public versions can be cached.
+            self.cache_request = True
+
+        log.bind(cache_request=self.cache_request)
+        log.debug("Serving docs.")
+
+        # Verify if the project is marked as spam and return a 401 in that case
+        spam_response = self._spam_response(request, project)
+        if spam_response:
+            return spam_response
+
+        # Check for canonical redirects before serving the file.
+        redirect_type = self._get_canonical_redirect_type(request)
+        if redirect_type:
+            try:
+                return self.canonical_redirect(
+                    request=request,
+                    final_project=project,
+                    version_slug=version.slug,
+                    filename=filename,
+                    redirect_type=redirect_type,
+                    is_external_version=unresolved.external,
+                )
+            except InfiniteRedirectException:
+                # Don't redirect in this case, since it would break things.
+                pass
```

Comment:
This code was already under the try/except clause. Why are we handling canonical redirects here as well? What are the different cases that each of them are handling?

Reply:
Our current implementation always does canonical redirects before checking for any other redirects or serving files, so we need it at both places, this won't be required with https://github.com/readthedocs/readthedocs.org/issues/10065.